File: poc-stt/geocodage_ign.py
"""Géocodage via la Géoplateforme IGN (data.geopf.fr) — API publique, sans clé.

Portage serveur de `frontend/src/data/geocodageIgn.ts`. Utilisé comme outil par
l'extraction LLM : Claude géocode une adresse avant de positionner une entité.
Règle produit : l'IA propose, l'humain valide — un score < SEUIL_FIABLE signale
une adresse « à confirmer ».
"""

import json
import urllib.parse
import urllib.request

SEUIL_FIABLE = 0.8


def geocoder(adresse: str) -> dict | None:
    """Géocode une adresse en coordonnées lon/lat.

    Returns:
        {lon, lat, score, label, fiable} pour le meilleur résultat, ou None si
        introuvable.
    """
    params = urllib.parse.urlencode({"q": adresse, "index": "address", "limit": "1"})
    url = f"https://data.geopf.fr/geocodage/search?{params}"

    with urllib.request.urlopen(url, timeout=15) as resp:  # noqa: S310 (API IGN publique)
        data = json.loads(resp.read())

    feature = (data.get("features") or [None])[0]
    coords = (feature or {}).get("geometry", {}).get("coordinates")
    if not feature or not coords:
        return None

    props = feature.get("properties", {})
    score = props.get("score", 0) or 0
    return {
        "lon": coords[0],
        "lat": coords[1],
        "score": score,
        "label": props.get("label", adresse),
        "fiable": score >= SEUIL_FIABLE,
    }

File: frontend/src/data/geocodageIgn.ts
// Géocodage via la Géoplateforme IGN (data.geopf.fr) — API publique, sans clé.
// Règle produit : l'IA propose, l'humain valide. Un score < SEUIL_FIABLE
// signale une adresse « à confirmer » (bandeau + validation manuelle).

/** En dessous, l'adresse est jugée incertaine → validation humaine requise. */
export const SEUIL_FIABLE = 0.8

export interface ResultatGeocodage {
  lon: number
  lat: number
  /** Score de confiance IGN, 0 → 1. */
  score: number
  /** Libellé normalisé renvoyé par l'IGN. */
  label: string
  /** true si `score >= SEUIL_FIABLE`. */
  fiable: boolean
}

interface FeatureIgn {
  geometry?: { coordinates?: [number, number] }
  properties?: { score?: number; label?: string }
}

/**
 * Géocode une adresse en coordonnées lon/lat.
 *
 * Args:
 *   adresse: Adresse en texte libre (ex. "12 rue des Lilas, Lyon").
 *
 * Returns:
 *   Le meilleur résultat, ou null si l'adresse est introuvable.
 */
export async function geocoder(adresse: string): Promise<ResultatGeocodage | null> {
  const url = new URL('https://data.geopf.fr/geocodage/search')
  url.searchParams.set('q', adresse)
  url.searchParams.set('index', 'address')
  url.searchParams.set('limit', '1')

  const res = await fetch(url)
  if (!res.ok) {
    throw new Error(`Géocodage IGN indisponible (HTTP ${res.status}).`)
  }
  const data = (await res.json()) as { features?: FeatureIgn[] }
  const feature = data.features?.[0]
  const coords = feature?.geometry?.coordinates
  if (!feature || !coords) return null

  const score = feature.properties?.score ?? 0
  return {
    lon: coords[0],
    lat: coords[1],
    score,
    label: feature.properties?.label ?? adresse,
    fiable: score >= SEUIL_FIABLE,
  }
}
